src/Microsoft.Identity.Client/Azure/ChainedTokenProvider.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Azure
+{
+    /// <summary>
+    /// ChainedTokenProvider creates an ordered list of probes that will determine if the environment will allow it to build
+    /// a ICredentialProvider. The ICredentialProvider is then used to generate authentication credentials for the
+    /// consumer.
+    /// </summary>
+    public class ChainedTokenProvider : ITokenProvider
+    {
+        private readonly IEnumerable<IProbe> _probes;
+
+        /// <summary>
+        /// Create an instance of a ChainedTokenProvider providing a list of IProbes which will be executed in order to create a ICredentialProvider
+        /// </summary>
+        /// <param name="probes">probes to be executed in order to create a ICredentialProvider</param>
+        /// <exception cref="ArgumentException">throws if no probes were provided or if no probe is able to build a ICredentialProvider</exception>
+        public ChainedTokenProvider(IEnumerable<IProbe> probes)
+        {
+            if (probes == null || probes.Count() < 1)
```

Comment:
Having probes be pre-collapsed (or collapsing it upon receipt) to an `IList<T>` or an `ICollection<T>` would be preferable here as it avoids issues that could arise from multiple enumeration. Collapsing it upon receipt would avoid collection modified issues when enumerating the elements. If maintaining `probes` being an `IEnumerable<T>` is desirable, `probes.Count() < 1` should be changed to `!probes.Any()` to avoid enumerating the entire set to determine whether there are zero elements.